# Rice_eccDNA: Analytical Code for Rice's Extrachromosomal Circular DNA Study

Welcome to the `Rice_eccDNA` repository. This repository contains the analytical code employed in the research paper titled "The Adaptational Dynamics of Extrachromosomal Circular DNA in Rice."

## Overview

This study offers insights into the role and significance of extrachromosomal circular DNA (eccDNA) within the rice genome. Our findings provide a deeper understanding of genome dynamics, stability, and open avenues for exploring rice cultivation strategies.

## Repository Contents

- **01_MultiFastqSampler.sh**: Efficiently sample multiple FASTQ files.
- **02_eccDNA_DualAnalysis.sh**: Streamline the identification and analysis of eccDNA.
- **03_Count_reads_in_bed_regions.py**: Count overlapping reads from a SAM file with a BED file.
- **04_Repeats_identification.py**: Identify direct repeats crucial for eccDNA formation.
- **Merge.R**: Merge and post-process identified eccDNAs.
- **genomeSize.py**: Analyze the genome's size relative to eccDNA abundance.

## Quick Start

To use the scripts provided, ensure you have all necessary dependencies installed. Each script contains comments that provide detailed usage instructions.

## Feedback and Contributions

Feedback, bug reports, and contributions are heartily welcomed. If you make use of this code in your research, please cite our original paper.

## License

This project is licensed under the [GNU General Public License v3.0](LICENSE). Kindly refer to the `LICENSE` file for detailed terms and conditions.
